server/channels/app/custom_profile_attributes_test.go

```diff
@@ -0,0 +1,462 @@
+}
+
+func TestGetCPAValue(t *testing.T) {
+	os.Setenv("MM_FEATUREFLAGS_CUSTOMPROFILEATTRIBUTES", "true")
+	defer os.Unsetenv("MM_FEATUREFLAGS_CUSTOMPROFILEATTRIBUTES")
+	th := Setup(t).InitBasic()
+	defer th.TearDown()
+
+	if cpaGroupID == "" {
+		_, err := th.App.cpaGroupID()
+		require.NoError(t, err)
+	}
+	fieldID := model.NewId()
+
+	t.Run("should fail if the value doesn't exist", func(t *testing.T) {
+		pv, appErr := th.App.GetCPAValue(model.NewId())
+		require.NotNil(t, appErr)
+		require.Nil(t, pv)
+	})
+
+	t.Run("should fail if the group id is invalid", func(t *testing.T) {
+		propertyValue := &model.PropertyValue{
+			TargetID:   model.NewId(),
+			TargetType: "user",
+			GroupID:    model.NewId(),
+			FieldID:    fieldID,
+			Value:      "Value",
+		}
+		propertyValue, err := th.App.Srv().propertyService.CreatePropertyValue(propertyValue)
+		require.NoError(t, err)
+
+		pv, appErr := th.App.GetCPAValue(propertyValue.ID)
+		require.NotNil(t, appErr)
+		require.Nil(t, pv)
+	})
+
+	t.Run("should succeed if id exists", func(t *testing.T) {
+		propertyValue := &model.PropertyValue{
+			TargetID:   model.NewId(),
+			TargetType: "user",
+			GroupID:    cpaGroupID,
+			FieldID:    fieldID,
+			Value:      "Value",
+		}
+		propertyValue, err := th.App.Srv().propertyService.CreatePropertyValue(propertyValue)
+		require.NoError(t, err)
+
+		pv, appErr := th.App.GetCPAValue(propertyValue.ID)
+		require.Nil(t, appErr)
+		require.NotNil(t, pv)
+	})
+}
+
+func TestPatchCPAValue(t *testing.T) {
+	os.Setenv("MM_FEATUREFLAGS_CUSTOMPROFILEATTRIBUTES", "true")
+	defer os.Unsetenv("MM_FEATUREFLAGS_CUSTOMPROFILEATTRIBUTES")
+	th := Setup(t).InitBasic()
+	defer th.TearDown()
+
+	if cpaGroupID == "" {
+		_, err := th.App.cpaGroupID()
+		require.NoError(t, err)
+	}
+
+	t.Run("should fail if the field doesn't exist", func(t *testing.T) {
+		invalidFieldID := model.NewId()
+		_, appErr := th.App.PatchCPAValue(model.NewId(), invalidFieldID, "fieldValue")
+		require.NotNil(t, appErr)
+	})
+
+	t.Run("should create value if new field value", func(t *testing.T) {
+		newField := &model.PropertyField{
+			GroupID: cpaGroupID,
+			Name:    model.NewId(),
+			Type:    model.PropertyFieldTypeText,
+		}
+		createdField, err := th.App.Srv().propertyService.CreatePropertyField(newField)
+		require.NoError(t, err)
+
+		userID := model.NewId()
+		patchedValue, appErr := th.App.PatchCPAValue(userID, createdField.ID, "test value")
+		require.Nil(t, appErr)
+		require.NotNil(t, patchedValue)
+		require.Equal(t, "test value", patchedValue.Value)
+		require.Equal(t, userID, patchedValue.TargetID)
+
+		t.Run("should correctly patch the CPA property value", func(t *testing.T) {
+			patch2, appErr := th.App.PatchCPAValue(userID, createdField.ID, "new patched value")
+			require.Nil(t, appErr)
+			require.NotNil(t, patch2)
+			require.Equal(t, patchedValue.ID, patch2.ID)
+			require.Equal(t, "new patched value", patch2.Value)
+			require.Equal(t, userID, patch2.TargetID)
+		})
+	})
```

Comment:
This I really like! A test being included inside another one, the dependency is clear.

Reply:
Heh, funnily enough, tackled your previous comment through nesting as well, glad it is clear that way :)